src/Hosting/Hosting/src/Internal/ServiceFactoryAdapter.cs

```diff
@@ -9,20 +9,42 @@ namespace Microsoft.Extensions.Hosting.Internal
     internal class ServiceFactoryAdapter<TContainerBuilder> : IServiceFactoryAdapter
     {
         private IServiceProviderFactory<TContainerBuilder> _serviceProviderFactory;
+        private readonly Func<HostBuilderContext> _contextResolver;
+        private Func<HostBuilderContext, IServiceProviderFactory<TContainerBuilder>> _factoryResolver;
 
         public ServiceFactoryAdapter(IServiceProviderFactory<TContainerBuilder> serviceProviderFactory)
         {
-            _serviceProviderFactory = serviceProviderFactory ?? throw new System.ArgumentNullException(nameof(serviceProviderFactory));
+            _serviceProviderFactory = serviceProviderFactory ?? throw new ArgumentNullException(nameof(serviceProviderFactory));
+        }
+
+        public ServiceFactoryAdapter(Func<HostBuilderContext> contextResolver, Func<HostBuilderContext, IServiceProviderFactory<TContainerBuilder>> factoryResolver)
```

Comment:
Why a func for the HostBuilderContext? Why not just pass it directly?

Reply:
Because HostBuilderContext isn't created until you call Build and this gets called first.

Reply:
I see.